storage/src/store/mod.rs

```diff
@@ -0,0 +1,986 @@
+where
+    E: RStorage + Clock + Metrics,
+    K: Array,
+    V: Span,
+    T: Translator,
+{
+    /// Initializes a new [`Store`] database with the given configuration.
+    ///
+    /// ## Rollback
+    ///
+    /// Any uncommitted operations will be rolled back if the [`Store`] was previously closed without
+    /// committing.
+    pub async fn init(
+        context: E,
+        cfg: Config<T, <Operation<K, V> as Read>::Cfg>,
+    ) -> Result<Self, Error> {
+        let snapshot: Index<T, u64> = Index::init(context.with_label("snapshot"), cfg.translator);
+
+        let log = VJournal::init(
+            context.with_label("log"),
+            VConfig {
+                partition: cfg.log_journal_partition,
+                compression: cfg.log_compression,
+                codec_config: cfg.log_codec_config,
+                write_buffer: cfg.log_write_buffer,
+            },
+        )
+        .await?;
+
+        let locations = FJournal::init(
+            context.with_label("locations"),
+            FConfig {
+                partition: cfg.locations_journal_partition,
+                items_per_blob: cfg.locations_items_per_blob,
+                write_buffer: cfg.log_write_buffer,
+                buffer_pool: cfg.buffer_pool.clone(),
+            },
+        )
+        .await?;
+
+        let db = Self {
+            log,
+            snapshot,
+            log_items_per_section: cfg.log_items_per_section,
+            locations,
+            inactivity_floor_loc: 0,
+            oldest_retained_loc: 0,
+            op_count: 0,
+            uncommitted_ops: 0,
+        };
+        db.build_snapshot_from_log().await
+    }
+
+    /// Gets the value associated with the given key in the store.
+    ///
+    /// If the key does not exist, returns `Ok(None)`.
+    pub async fn get(&self, key: &K) -> Result<Option<V>, Error> {
+        for location in self.snapshot.get(key) {
+            let (k, v) = Self::get_set_op(
+                &self.log,
+                &self.locations,
+                self.log_items_per_section,
+                *location,
+            )
+            .await?;
+
+            if &k == key {
+                return Ok(Some(v));
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Updates the value associated with the given key in the store.
+    ///
+    /// If the key already has the same value, this is a no-op and no log entry is created.
+    ///
+    /// The operation is immediately visible in the snapshot for subsequent queries, but remains
+    /// uncommitted until [`Store::commit`] is called. Uncommitted operations will be rolled back
+    /// if the store is closed without committing.
+    pub async fn update(&mut self, key: K, value: V) -> Result<UpdateResult, Error> {
+        // Check if this is a no-op
+        if let Some(current_value) = self.get(&key).await? {
+            if current_value == value {
+                return Ok(UpdateResult::NoOp);
+            }
+        }
```

Comment:
Curious y'all's opinion on whether or not this is desired. Removing this idea of no-op'ing `update` calls when the key has already assumed the same value results in a slightly more expensive pruning process due to unconditionally appending of a new operation, but removes an extra read from the locations map + op log in the `update` path.

Reply:
Should get around to reviewing this sometime tomorrow 👍